domain/secret/service/watcher.go

```diff
@@ -155,6 +150,237 @@ func (s *WatchableService) WatchSecretRevisionsExpiryChanges(_ context.Context,
+	watcherFactory WatcherFactory,
+	logger logger.Logger,
+	owners ...CharmSecretOwner,
+) (*obsoleteWatcher, error) {
+	if len(owners) == 0 {
+		return nil, internalerrors.Errorf("at least one owner must be provided for the obsolete secret watcher")
+	}
+
+	appOwners, unitOwners := splitCharmSecretOwners(owners...)
+	w := &obsoleteWatcher{
+		logger:          logger,
+		state:           state,
+		watcherFactory:  watcherFactory,
+		appOwners:       appOwners,
+		unitOwners:      unitOwners,
+		knownSecretURIs: set.NewStrings(),
+		out:             make(chan []string),
+	}
+
+	if err := w.init(); err != nil {
+		return nil, internalerrors.Capture(err)
+	}
+
+	err := catacomb.Invoke(catacomb.Plan{
+		Site: &w.catacomb,
+		Work: w.loop,
+		Init: []worker.Worker{w.secretWatcher, w.revisionWatcher},
+	})
+	return w, internalerrors.Capture(err)
+}
+
+func (w *obsoleteWatcher) init() error {
+	var err error
+	table, query := w.state.InitialWatchStatementForOwnedSecrets(w.appOwners, w.unitOwners)
+	w.secretWatcher, err = w.watcherFactory.NewNamespaceWatcher(
+		table,
+		changestream.All,
+		query,
+	)
+	if err != nil {
+		return internalerrors.Capture(err)
+	}
+
+	table, query = w.state.InitialWatchStatementForObsoleteRevision(w.appOwners, w.unitOwners)
+	w.revisionWatcher, err = w.watcherFactory.NewNamespaceWatcher(
+		table,
+		changestream.Changed,
+		query,
+	)
+	return internalerrors.Capture(err)
+}
+
+func (w *obsoleteWatcher) scopedContext() (context.Context, context.CancelFunc) {
+	return context.WithCancel(w.catacomb.Context(context.Background()))
+}
+
+func splitSecretRevision(s string) (string, int) {
+	parts := strings.Split(s, "/")
+	if len(parts) < 2 {
+		return parts[0], 0
+	}
+	rev, _ := strconv.Atoi(parts[1])
+	return parts[0], rev
+}
+
+func (w *obsoleteWatcher) mergeSecretChanges(
+	ctx context.Context, currentChanges []string, receivcedSecretIDs []string,
+) ([]string, error) {
+	if len(receivcedSecretIDs) == 0 {
+		return currentChanges, nil
+	}
+
+	// pushChanges pushes the secret ID to the changes slice.
+	// At the same time, any previously added obsolete revisions of this secret are removed from the slice.
+	pushChanges := func(secretID string) {
+		currentChanges = slices.DeleteFunc(currentChanges, func(s string) bool {
+			id, _ := splitSecretRevision(s)
+			return id == secretID
+		})
+		currentChanges = append(currentChanges, secretID)
+	}
+
+	w.logger.Criticalf(ctx, "secretWatcher(%v-%v): received changes %v", w.appOwners, w.unitOwners, receivcedSecretIDs)
+	for _, uriStr := range receivcedSecretIDs {
+		uri, err := coresecrets.ParseURI(uriStr)
+		if err != nil {
+			return currentChanges, errors.Trace(err)
+		}
+		owned, err := w.state.IsSecretOwnedBy(ctx, uri, w.appOwners, w.unitOwners)
+		w.logger.Criticalf(ctx, "secretWatcher(%v-%v): uri %q, owned %v", w.appOwners, w.unitOwners, uri, owned)
+		if err != nil && !errors.Is(err, secreterrors.SecretNotFound) {
+			return currentChanges, errors.Trace(err)
+		}
+		if owned {
```

Comment:
Need to account for the SecretNotFoundError. owned will be false but we still want to be explicit about handling the error. Just add it to the check below.

Reply:
After discussed, we can treat not found == not owned here.